src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
-        private async ValueTask AcquireWriteLockAsync()
+        private async ValueTask StartWriteAsync(int writeBytes, CancellationToken cancellationToken = default, bool acquireWriteLock = true)
         {
-            await _writerLock.WaitAsync().ConfigureAwait(false);
-
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (IsAborted())
+            if (acquireWriteLock)
             {
-                throw new IOException(SR.net_http_invalid_response);
+                await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
             }
-        }
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
```

Comment:
Where did this 32768 number come from?  Is that tied to something else?  Configurable?  Assuming it's something we can dial and/or that it's tied to something else in the code, we should make it a const at the top of the class.  You might also consider defining it as `32 * 1024` to make it clear what the number is (though most folks reading the code will probably recognize it).

Reply:
It's pretty much just an arbitrary value. I'll move it to the top of the class, and add a comment describing what it means.